### PR TITLE
ARROW-6134: [C++][Gandiva] Add concat function in Gandiva

### PR DESCRIPTION
**Pull request by pprudhvi**
- remove concat alias for concatOperator
- add concat(utf8, utf8) function. The difference between concat and concatOperator is in null input handling. concatOperator returns null if one of the inputs is null; concat treats null input as empty string

### CODE REVIEW

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 1]
```diff
@@ -231,10 +231,28 @@ char* substr_utf8_int64(int64 context, const char* input, int32 in_len, int64 of
...
+char* concat_utf8_utf8(int64 context, const char* left, int32 left_len,
+                       bool left_validity, const char* right, int32 right_len,
+                       bool right_validity, bool* out_valid, int32* out_len) {
+  if (!left_validity) {
+    left_len = 0;
+  }
+  if (!right_validity) {
+    right_len = 0;
+  }
+  *out_valid = true;
```
Comment (praveenbingo) [inline cpp/src/gandiva/precompiled/string_ops.cc]: is this a null never function? since the output is set as always valid? (but function says null internal)..

can we please add a test case too.

Comment (pprudhvi) [reply cpp/src/gandiva/precompiled/string_ops.cc]: done

**Review by praveenbingo — APPROVED** [review]
looks good to me. Will merge once tests are green.

File: cpp/src/gandiva/function_registry_string.cc [hunk 2]
```diff
@@ -79,10 +79,13 @@ std::vector<NativeFunction> GetStringFunctionRegistry() {
                      utf8(), kResultNullIfNull, "substr_utf8_int64",
                      NativeFunction::kNeedsContext),
 
-      NativeFunction("concatOperator", {"concat"}, DataTypeVector{utf8(), utf8()}, utf8(),
+      NativeFunction("concatOperator", {}, DataTypeVector{utf8(), utf8()}, utf8(),
```
Comment (pravindra) [inline cpp/src/gandiva/function_registry_string.cc]: can you please add a comment here on the difference between the two ?

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 3]
```diff
@@ -207,7 +209,8 @@ char* substr_utf8_int64_int64(int64 context, const char* input, int32 in_len,
 
   if (startIndex < 0 || startIndex >= in_len) {
     *out_len = 0;
-    return nullptr;
+    const char* empty_str = "";
```
[the remark this answers is not in the document]
Comment (pprudhvi) [reply cpp/src/gandiva/precompiled/string_ops.cc]: done

File: cpp/src/gandiva/function_registry_string.cc [hunk 4]
```diff
@@ -79,10 +79,15 @@ std::vector<NativeFunction> GetStringFunctionRegistry() {
...
                      kResultNullIfNull, "concatOperator_utf8_utf8",
                      NativeFunction::kNeedsContext),
 
+      // concat treats null inputs as empty strings whereas concatOperator returns null if
+      // one of the inputs is null
+      NativeFunction("concat", {}, DataTypeVector{utf8(), utf8()}, utf8(),
```
Comment (bkietz) [inline cpp/src/gandiva/function_registry_string.cc]: I think it would be more clear to name these concat and concatNullAsEmpty or similar.

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 5]
```diff
@@ -190,11 +191,12 @@ VAR_LEN_TYPES(IS_NOT_NULL, isnotnull)
...
 FORCE_INLINE
-char* substr_utf8_int64_int64(int64 context, const char* input, int32 in_len,
-                              int64 offset64, int64 length, int32* out_len) {
+const char* substr_utf8_int64_int64(int64 context, const char* input, int32 in_len,
+                                    int64 offset64, int64 length, int32* out_len) {
   if (length <= 0 || input == nullptr || in_len <= 0) {
     *out_len = 0;
-    return nullptr;
+    const char* empty_str = "";
+    return const_cast<char*>(empty_str);
```
Comment (bkietz) [inline cpp/src/gandiva/precompiled/string_ops.cc]: do you need this cast anymore now that the function returns `const char*`?

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 6]
```diff
@@ -165,8 +165,9 @@ FORCE_INLINE
 char* castVARCHAR_utf8_int64(int64 context, const char* data, int32 data_len,
                              int64_t out_len, int32_t* out_length) {
   // TODO: handle allocation failures
-  int32_t len = data_len <= static_cast<int32_t>(out_len) ? data_len
-                                                          : static_cast<int32_t>(out_len);
+  int32_t len = data_len <= static_cast<int32_t>(out_len) || out_len == 0
```
Comment (bkietz) [inline cpp/src/gandiva/precompiled/string_ops.cc]: What is the significance of this change? It seems unrelated to the concat functions and you haven't added a new test for `out_len == 0`

**Review by bkietz — CHANGES_REQUESTED** [review]

File: cpp/src/gandiva/precompiled/string_ops.cc (hunk 6 above)
Comment (pprudhvi) [reply cpp/src/gandiva/precompiled/string_ops.cc]: this is not related to concat directly. I will create a new issue for this.

File: cpp/src/gandiva/precompiled/string_ops.cc (hunk 5 above)
Comment (pprudhvi) [reply cpp/src/gandiva/precompiled/string_ops.cc]: changed. thanks for pointing out

File: cpp/src/gandiva/function_registry_string.cc (hunk 4 above)
Comment (pprudhvi) [reply cpp/src/gandiva/function_registry_string.cc]: the name should match sql function name

**Review by pravindra — APPROVED** [review]
